br/pkg/storage/gcs_extra.go

```diff
@@ -0,0 +1,428 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Learned from https://github.com/liqiuqing/gcsmpu
+
+package storage
+
+import (
+	"bytes"
+	"context"
+	"encoding/xml"
+	"fmt"
+	"net"
+	"net/http"
+	"net/url"
+	"runtime"
+	"slices"
+	"strconv"
+	"sync"
+	"time"
+
+	"cloud.google.com/go/storage"
+	"github.com/go-resty/resty/v2"
+	"go.uber.org/atomic"
+)
+
+// GCSWriter uses XML multipart upload API to upload a single file.
+// https://cloud.google.com/storage/docs/multipart-uploads.
+// GCSWriter will attempt to cancel uploads that fail due to an exception.
+// If the upload fails in a way that precludes cancellation, such as a
+// hardware failure, process termination, or power outage, then the incomplete
+// upload may persist indefinitely. To mitigate this, set the
+// `AbortIncompleteMultipartUpload` with a nonzero `Age` in bucket lifecycle
+// rules, or refer to the XML API documentation linked above to learn more
+// about how to list and delete individual downloads.
+type GCSWriter struct {
+	uploadBase
+	mutex       sync.Mutex
+	xmlMPUParts []*xmlMPUPart
+	wg          sync.WaitGroup
+	err         atomic.Error
+	chunkSize   int64
+	workers     int
+	totalSize   int64
+	// only available for input reader which implements io.Seeker
+	readerPos int64
+	uploadID  string
+	chunkCh   chan chunk
+	curPart   int
+}
+
+// NewGCSWriter returns a GCSWriter which uses GCS multipart upload API behind the scene.
+func NewGCSWriter(
+	ctx context.Context,
+	cli *storage.Client,
+	uri string,
+	partSize int64,
+	parallelCnt int,
+	bucketName string,
+) (*GCSWriter, error) {
+	if partSize < gcsMinimumChunkSize || partSize > gcsMaximumChunkSize {
+		return nil, fmt.Errorf(
+			"invalid chunk size: %d. Chunk size must be between %d and %d",
+			partSize, gcsMinimumChunkSize, gcsMaximumChunkSize,
+		)
+	}
+
+	w := &GCSWriter{
+		uploadBase: uploadBase{
+			ctx:             ctx,
+			cli:             cli,
+			bucket:          bucketName,
+			blob:            uri,
+			retry:           defaultRetry,
+			signedURLExpiry: defaultSignedURLExpiry,
+		},
+		chunkSize: partSize,
+		workers:   parallelCnt,
+	}
+	if err := w.init(); err != nil {
+		return nil, fmt.Errorf("failed to initiate GCSWriter: %w", err)
+	}
+
+	return w, nil
+}
+
+func (w *GCSWriter) init() error {
+	opts := &storage.SignedURLOptions{
+		Scheme:          storage.SigningSchemeV4,
+		Method:          "POST",
+		Expires:         time.Now().Add(w.signedURLExpiry),
+		QueryParameters: url.Values{mpuInitiateQuery: []string{""}},
+	}
+	u, err := w.cli.Bucket(w.bucket).SignedURL(w.blob, opts)
+	if err != nil {
+		return fmt.Errorf("Bucket(%q).SignedURL: %s", w.bucket, err)
+	}
+
+	client := resty.New()
+	resp, err := client.R().Post(u)
+	if err != nil {
+		return fmt.Errorf("POST request failed: %s", err)
+	}
```

Comment:
How many retry times of the client?

Reply:
Seems not retry internally, it only rely on upper layer retry